### PR TITLE
feat: modified click behavior of auro-popover

### PR DESCRIPTION
**Pull request by BrentLabasan**
# Alaska Airlines Pull Request

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

**Fixes:** https://github.com/AlaskaAirlines/auro-popover/issues/6

## Summary:
Modified behavior of clicks events of auro-popover, so that it can work with 606028 https://dev.azure.com/itsals/E_Sell_Shop/_workitems/edit/606028

_Please summarize the scope of the changes you have submitted, what the intent of the work is and anything that describes the before/after state of the project._

## Type of change:

Please delete options that are not relevant.

- [ ] New capability
- [x] Revision of an existing capability
- [ ] Infrastructure change (automation, etc.)
- [ ] Other (please elaborate)

## Checklist:

- [x] My update follows the CONTRIBUTING guidelines of this project
- [x] I have performed a self-review of my own update

**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._

**Thank you for your submission!**<br>
-- Auro Design System Team

### CODE REVIEW

File: demo/demo.md [hunk 1]
```diff
@@ -22,39 +22,39 @@ In these examples the structure of the HTML uses the slot element to pull the tr
 
 <div class="exampleWrapper">
   <auro-popover for="button1">
-    Top popover content!
+    <fs-content-non-combinability slot="tooltip"></fs-content-non-combinability>
```
Comment (gusnaughton) [inline demo/demo.md:25]:
```
<link rel="stylesheet" href="https://p2pcontent-fd-prod.azurefd.net/components/shared/component-base.css">
<script src="https://p2pcontent-fd-prod.azurefd.net/components/shared/component-base.js" type="text/javascript"></script>
<script src="https://p2pcontent-fd-prod.azurefd.net/components/fs-content-non-combinability/main.js" type="text/javascript"></script>
```

File: demo/demo.md [hunk 2]
```diff
@@ -22,39 +22,39 @@ In these examples the structure of the HTML uses the slot element to pull the tr
...
     <auro-button id="button1" slot="trigger">Popover Test</auro-button>
   </auro-popover>
 
-  <auro-popover for="button2" placement="bottom">
+  <!-- <auro-popover for="button2" placement="bottom">
```
Comment (gusnaughton) [inline demo/demo.md:29]: Please uncomment the rest of the demos

File: src/auro-popover.js [hunk 3]
```diff
@@ -11,6 +11,9 @@ import styleCss from "./style-css.js";
 
 import Popover from "./popover";
 
+// I need this import so that the index.html/demo.md renders the popover correctly. I don't know exactly why.
```
Comment (gusnaughton) [inline src/auro-popover.js:14]: The above suggestion removes this dependency.

File: src/nonCombinability/index.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
+import {css, html, LitElement, unsafeCSS} from 'lit-element';
+// import {isIE11} from '../../shared/utility';
+// import '../../Generics/import-auro-components';
```
Comment (gusnaughton) [inline src/nonCombinability/index.js:3]: Unused file now, right?

File: src/nonCombinability/demo.html [hunk 5]
```diff
@@ -0,0 +1,7 @@
+<script>
```
Comment (gusnaughton) [inline src/nonCombinability/demo.html:1]: Unused file

File: src/nonCombinability/styles.css [hunk 6]
```diff
@@ -0,0 +1,17 @@
+.tooltip {
```
Comment (gusnaughton) [inline src/nonCombinability/styles.css:1]: Unused file

File: src/nonCombinability/styles.scss [hunk 7]
```diff
@@ -0,0 +1,26 @@
+.tooltip{
```
Comment (gusnaughton) [inline src/nonCombinability/styles.scss:1]: Everything in src/nonCombinability will be unused

**Review by gusnaughton — CHANGES_REQUESTED** [review]

File: src/popover.js [hunk 8]
```diff
@@ -20,8 +20,6 @@ export default class Popover {
     }
 
     show() {
-        this.popover.hidden = false;
```
Comment (BrentLabasan) [inline src/popover.js:23]: unused variable

File: demo/demo.md (hunk 1 above)
Comment (gusnaughton) [reply demo/demo.md]: @blackfalcon it seems like adding the above lines does not actually cause the browser to render them in the new markdown viewer. Is this a known limitation? Do you know a workaround?

Comment (blackfalcon) [reply demo/demo.md]: The additional script tags need to be added to the index.html file.

File: README.md [hunk 9]
```diff
@@ -108,5 +108,9 @@ $ polymer serve
 
 Open [localhost:8000](http://localhost:8000/)
 
+### Commit message formats
```
Comment (blackfalcon) [inline README.md:111]: This is not needed here. 

A full spec for how to manage feature branches and pull requests can be found here. 
https://auro.alaskaair.com/getting-started/developers/contributing

Please remove this commit.

**Comment by blackfalcon** [comment]
@gusnaughton @BrentLabasan please let me know when you are at a good stopping place for me to do a code review.

**Comment by gusnaughton** [comment]
I modified the demo to work with a internal toggle() function on the Popover class. @BrentLabasan, does this achieve the intended end result?

File: README.md (hunk 9 above)
Comment (BrentLabasan) [reply README.md:111]: Could that documentation then include the notice that their local commits on their machine won't go through unless they also conform to the standard?

File: package.json [hunk 10]
```diff
@@ -102,7 +102,7 @@
     "postinstall": "node packageScripts/postinstall.js",
     "postinstall:lit": "node scripts/postinstall.js",
     "ciBuild": "npm-run-all sassBuild sassRender cssLint distJS bundler postinstall apiBuild",
-    "cssLint": "stylelint './src/*.css'",
+    "cssLint": "stylelint ./src/*.css",
```
Comment (BrentLabasan) [inline package.json:105]: This modification allows the command to run without error on my Windows machine, which is a hook in the commit process.

Comment (gusnaughton) [reply package.json:105]: Please submit this to the WC-Generator repo as well! this is a great find.

Comment (blackfalcon) [reply package.json:105]: This is how we have this in the current version of the Generator. 
https://github.com/AlaskaAirlines/WC-Generator/blob/master/template/package.json#L101

When we ran into this in the past, that was the preferred solution.

File: README.md [hunk 11]
```diff
@@ -101,6 +101,9 @@ $ npm run dev
 
 // shell terminal two
 $ npm run serve
+
+// shell terminal three
```
Comment (blackfalcon) [inline README.md:105]: Why is this added here? There are no polymer dependencies in this repo?

File: src/auro-popover.js [hunk 12]
```diff
@@ -115,7 +162,7 @@ class AuroPopover extends LitElement {
     return html`
       <div id="popover" class="popover util_insetLg">
         <div id="arrow" class="arrow" data-popper-arrow></div>
-        <slot role="tooltip"></slot>
+        <slot name="tooltip"></slot>
```
Comment (blackfalcon) [inline src/auro-popover.js:165]: Why was this changed? 

The purpose of `role` is for accessibility. This also made the use of the component require to have to use a named slot whereas the generic slot would have populated this element prior.

File: demo/demo.md [hunk 13]
```diff
@@ -22,17 +22,24 @@ In these examples the structure of the HTML uses the slot element to pull the tr
 
 <div class="exampleWrapper">
   <auro-popover for="button1">
-    Top popover content!
+    <div slot="tooltip">Top popover content!</div>
```
Comment (blackfalcon) [inline demo/demo.md:25]: This is not required. 

Return the template to use `role` and then there is no requirement to wrap this string in a slot reference.

File: src/auro-popover.js [hunk 14]
```diff
@@ -16,17 +16,22 @@ import Popover from "./popover";
...
+ * @attr {boolean} isPopoverVisible - Boolean for if popover is visible or not.
+ * @attr {boolean} isSticky - If true, when user clicks trigger, the popover will persist its visibility. If false, popover will disappear when mouseout over trigger.
  * @slot - Default unnamed slot for the use of popover content
  * @slot trigger - Slot for entering the trigger element into the scope of the shadow DOM
  */
 class AuroPopover extends LitElement {
   constructor() {
     super();
     this.placement = 'top';
+    this.isPopoverVisible = false;
```
Comment (blackfalcon) [inline src/auro-popover.js:28]: This is part of the private API and should not be exposed in the docs. 

See this code as an example of how to remove this from the generated API doc
https://github.com/AlaskaAirlines/auro-icon/blob/master/src/auro-icon.js

File: docs/api.md [hunk 15]
```diff
@@ -4,10 +4,18 @@ Popover attaches to an element and displays on hover/blur.
...
+| `for`              | `for`              | `String`  |         | Defines an `id` for an element in the DOM to trigger on hover/blur. |
+| `isPopoverVisible` | `isPopoverVisible` | `boolean` | false   | Boolean for if popover is visible or not.        |
+| `isSticky`         | `isSticky`         | `boolean` |         | If true, when user clicks trigger, the popover will persist its visibility. If false, popover will disappear when mouseout over trigger. |
+| `placement`        | `placement`        | `String`  | "top"   | Expects top/bottom - position for popover in relation to the element. |
+
+## Methods
+
+| Method   | Type       | Description                      |
+|----------|------------|----------------------------------|
+| `toggle` | `(): Void` | Toggles the popover's open state |
```
Comment (blackfalcon) [inline docs/api.md:18]: This method also should be hidden from the public API docs.

File: src/auro-popover.js [hunk 16]
```diff
@@ -60,53 +65,95 @@ class AuroPopover extends LitElement {
...
-    });
+    // e.g. for a closePopover button in the popover
+    this.addEventListener('hidePopover', toggleHide);
 
+    // if user clicks on something other than trigger or popover, close popover
+    document.addEventListener('click', handleClickNonTriggerNonPopover);
+  }
+
+  /**
+    * Toggles the popover's open state
```
Comment (blackfalcon) [inline src/auro-popover.js:144]:
```suggestion
    * @private Toggles the popover's open state
```

File: src/auro-popover.js [hunk 17]
```diff
@@ -16,17 +16,22 @@ import Popover from "./popover";
...
+ * @attr {boolean} isSticky - If true, when user clicks trigger, the popover will persist its visibility. If false, popover will disappear when mouseout over trigger.
  * @slot - Default unnamed slot for the use of popover content
  * @slot trigger - Slot for entering the trigger element into the scope of the shadow DOM
  */
 class AuroPopover extends LitElement {
   constructor() {
     super();
     this.placement = 'top';
+    this.isPopoverVisible = false;
+    this.isSticky = this.hasAttribute('sticky');
```
Comment (blackfalcon) [inline src/auro-popover.js:29]: This is not needed. https://lit-element.polymer-project.org/guide/properties#declare

Delete line.

File: src/auro-popover.js [hunk 18]
```diff
@@ -16,17 +16,22 @@ import Popover from "./popover";
...
   }
 
   // function to define props used within the scope of this component
   static get properties() {
+
     return {
       placement:  { type: String },
       for:        { type: String },
```
Comment (blackfalcon) [inline src/auro-popover.js:37]:
```suggestion
      for:        { type: String },
      sticky:   { type: String }
```

File: docs/api.md [hunk 19]
```diff
@@ -4,10 +4,18 @@ Popover attaches to an element and displays on hover/blur.
...
 
-| Property    | Attribute   | Type     | Default | Description                                      |
-|-------------|-------------|----------|---------|--------------------------------------------------|
-| `for`       | `for`       | `String` |         | Defines an `id` for an element in the DOM to trigger on hover/blur. |
-| `placement` | `placement` | `String` | "top"   | Expects top/bottom - position for popover in relation to the element. |
+| Property           | Attribute          | Type      | Default | Description                                      |
+|--------------------|--------------------|-----------|---------|--------------------------------------------------|
+| `for`              | `for`              | `String`  |         | Defines an `id` for an element in the DOM to trigger on hover/blur. |
+| `isPopoverVisible` | `isPopoverVisible` | `boolean` | false   | Boolean for if popover is visible or not.        |
+| `isSticky`         | `isSticky`         | `boolean` |         | If true, when user clicks trigger, the popover will persist its visibility. If false, popover will disappear when mouseout over trigger. |
```
Comment (blackfalcon) [inline docs/api.md:11]: This is really confusing. The actual API property is `sticky`. That is what the user will add to the element for this feature. 

The `isSticky` is internal.

File: src/auro-popover.js [hunk 20]
```diff
@@ -16,17 +16,22 @@ import Popover from "./popover";
  *
  * @attr {String} placement - Expects top/bottom - position for popover in relation to the element.
  * @attr {String} for - Defines an `id` for an element in the DOM to trigger on hover/blur.
+ * @attr {boolean} isPopoverVisible - Boolean for if popover is visible or not.
+ * @attr {boolean} isSticky - If true, when user clicks trigger, the popover will persist its visibility. If false, popover will disappear when mouseout over trigger.
```
Comment (blackfalcon) [inline src/auro-popover.js:20]: This is incorrect. The property a user will use is `sticky`. That is what you are using in your demo.

File: src/auro-popover.js [hunk 21]
```diff
@@ -60,53 +65,95 @@ class AuroPopover extends LitElement {
...
+       * @param {Event} event event
+       * @returns {Void} Fires an update lifecycle
+       */
+      handleTabWhenFocusOnTrigger = (event) => {
+        if (event.key.toLowerCase() === 'tab') {
+          toggleHide();
+        }
+      };
+
+    if (this.isSticky) {
```
Comment (blackfalcon) [inline src/auro-popover.js:125]:
```suggestion
    if (this.sticky) {
```

**Review by blackfalcon — CHANGES_REQUESTED** [review]
Pretty close. There are just some issues regarding internal versus external API. Also, I put some notes in there regarding a misunderstanding of the lit element API. 

Building custom elements, we can declare our own attributes. 

Please read https://lit-element.polymer-project.org/guide/properties#declare to understand more about attributes and properties in this context.

**Comment by blackfalcon** [comment]
Please update the demo. 

I'd prefer that the STICKY version is moved into its own block. That way it is clear that it is not part of the default experience.

**Comment by blackfalcon** [comment]
As a form of process. Please DO NOT merge `master` down into your feature branch. It is preferred that the feature branch is rebased on top of `master`.

**Comment by blackfalcon** [comment]
Looking for a status update on this PR. Removed `review` badge and replaced with `wip` badge.

When the PR is ready for review, please swap these badges and alert the reviewers that this PR is once again ready for review. 

Thanks!

File: README.md (hunk 11 above)
Comment (BrentLabasan) [reply README.md:105]: On line 94 it says to open three different shell sessions, with the third to "run the Polymer server".
But I tried running the demo.html without the third shell session running Polymer, and I can view the demo site.

So is the last shell session needed or not?

File: src/auro-popover.js (hunk 12 above)
Comment (BrentLabasan) [reply src/auro-popover.js]: Reverted.

File: demo/demo.md (hunk 13 above)
Comment (BrentLabasan) [reply demo/demo.md]: Reverted.

File: src/auro-popover.js (hunk 14 above)
Comment (BrentLabasan) [reply src/auro-popover.js:38]: Line has been changed according to the example provided.

File: src/auro-popover.js (hunk 16 above)
Comment (BrentLabasan) [reply src/auro-popover.js]: Did suggestion, in the code itself.

File: src/auro-popover.js (hunk 17 above)
Comment (BrentLabasan) [reply src/auro-popover.js]: Lines deleted.

File: src/auro-popover.js (hunk 18 above)
Comment (BrentLabasan) [reply src/auro-popover.js:45]: The _sticky:   { type: String }_ should be of type Boolean, right? Since configuring the popover to be sticky will be done with
<auro-popover **sticky**>

File: docs/api.md (hunk 19 above)
Comment (BrentLabasan) [reply docs/api.md]: Changed.

File: src/auro-popover.js (hunk 20 above)
Comment (BrentLabasan) [reply src/auro-popover.js]: Changed.

File: src/auro-popover.js (hunk 21 above)
Comment (BrentLabasan) [reply src/auro-popover.js]: Changed.

**Comment by BrentLabasan** [comment]
I want to switch my PR from Work In Progress to Review, but I don't see the checkbox next to the PR item to select it to change the label on it, as shown in https://docs.github.com/en/github/managing-your-work-on-github/managing-labels#applying-labels-to-issues-and-pull-requests

File: src/auro-popover.js [hunk 22]
```diff
@@ -60,53 +74,95 @@ class AuroPopover extends LitElement {
...
+       */
+      handleTabWhenFocusOnTrigger = (event) => {
+        if (event.key.toLowerCase() === 'tab') {
+          toggleHide();
+        }
+      };
+
+    if (this.sticky) {
+      trigger.addEventListener('click', toggleShow);
+      trigger.addEventListener('focus', toggleShow);
```
Comment (blackfalcon) [inline src/auro-popover.js:136]:
```suggestion

```

I am going to ask to remove this. As an accessibility attribute of the experience, things popping up when tabbed to them is not desired. Also, this is not consistent with the other popovers in the demo. 

The specification of the sticky popover is to have it stay open on a click versus a fly-by when hovered. 

Also, this update alerted me to the fact that our popovers are not really accessible. There is no way to tab into the scope of the popover. Also, the content within the popover is tabbable and ready aloud even when the popover is closed. 

These a11y issues will be addressed in another issue.

**Review by blackfalcon — CHANGES_REQUESTED** [review]
This is really good work. With these last changes, we should be able to release this feature.

File: docs/api.md (hunk 15 above)
Comment (blackfalcon) [reply docs/api.md]: Yeah, this was my bad. I didn't realize that this needed to be used in a public space.

**Review by blackfalcon — APPROVED** [review]
